### PR TITLE
refactor: port from github3.py to PyGithub

### PR DESCRIPTION
**Pull request by jmeridth**
## Proposed Changes

Port from `github3.py` to `PyGithub` for GitHub API interactions, aligning this repo with the same library used across other community projects (pr-conflict-detector, measure-innersource).

## What/Why

`github3.py` is being replaced with `PyGithub` across OSPO projects for consistency. PyGithub handles retry/timeout internally, removing the need for the `requests` dependency.

## Proof it works

73 tests pass with 100% coverage.

## Risk + AI role

Low -- 1:1 API migration with no behavioral changes. AI-assisted (Claude Opus 4.6) with human review.

## Review focus

- Auth flow in `auth.py` -- App auth now uses `Auth.AppAuth` + `get_installation_auth` instead of `login_as_app_installation`
- Large file blob handling in `cleanowners.py` -- base64 decode path via `get_git_blob`

### Author/Contributor

- [x] If documentation is needed for this change, has that been included in this pull request
- [x] run `make lint` and fix any issues that you have introduced
- [x] run `make test` and ensure you have test coverage for the lines you are introducing

### CODE REVIEW

File: cleanowners.py [hunk 1]
```diff
@@ -192,7 +193,7 @@ def main():  # pragma: no cover
                     break
 
                 # Check to see if the username is a member of the organization
-                if not gh_org.is_member(username):
+                if not gh_org.has_in_members(username):
```
Comment (zkoppert) [inline cleanowners.py]: This is the core membership check, and I believe it's currently broken for the main use case. PyGithub's `Organization.has_in_members` asserts `isinstance(member, github.NamedUser.NamedUser)` and raises `AssertionError` on a plain string, where github3's `is_member` accepted a login. I reproduced `AssertionError: octocat` in the venv, so the first individual `@handle` in any CODEOWNERS trips the outer `except Exception` and the whole run errors out. Resolving the login to a user object first fixes it:

```suggestion
                # Check to see if the username is a member of the organization
                member = github_connection.get_user(username)
                if not gh_org.has_in_members(member):
```

One edge case worth handling: `get_user` raises `UnknownObjectException` for a renamed or deleted login, which is arguably itself a signal that the user should be suggested for removal. The tests use `MagicMock` orgs so they pass today; a test that passes a real-typed org (or asserts a `NamedUser` reaches `has_in_members`) would lock this in.

**Review by zkoppert — COMMENTED** [review]
Multi-model review of the github3 to PyGithub migration below (Opus, GPT-5.3-Codex, GPT-5.4, all findings verified in the repo venv). Two verified Highs where the PyGithub object types differ from github3: `has_in_members` needs a `NamedUser` rather than a login string, and the large-file `content is None` guard never fires. I also flagged the exception-scope broadening in the read paths. Details inline.

File: cleanowners.py [hunk 2]
```diff
@@ -164,19 +165,19 @@ def main():  # pragma: no cover
...
-                codeowners_decoded = repo.blob(
-                    repo.file_contents(codeowners_filepath).sha
-                ).decode_content()
+                blob = repo.get_git_blob(repo.get_contents(codeowners_filepath).sha)
+                codeowners_decoded = blob.content.encode("utf-8")
+                if blob.encoding == "base64":
+                    codeowners_decoded = base64.b64decode(blob.content)
             else:
-                codeowners_decoded = codeowners_file_contents.decoded
+                codeowners_decoded = codeowners_file_contents.decoded_content
```
Comment (zkoppert) [inline cleanowners.py]: Two things here, both verified in the venv against a real large file. First, `codeowners_file_contents.content is None` is never true under PyGithub, so the new blob-download branch is dead code: github3 set `.content = None` for large files, but PyGithub returns `content == ""` with `encoding == "none"`. Second, because the guard is never true, a large CODEOWNERS falls through to `decoded_content`, which asserts `encoding == "base64"` and raises `AssertionError: unsupported encoding: none`, so the run errors out on exactly the case this branch was meant to handle. Gating on the encoding fixes both, and we can reuse the sha we already have instead of a second `get_contents` call:

```suggestion
            if codeowners_file_contents.encoding == "none":
                blob = repo.get_git_blob(codeowners_file_contents.sha)
                codeowners_decoded = base64.b64decode(blob.content)
            else:
                codeowners_decoded = codeowners_file_contents.decoded_content
```

A test that feeds a `ContentFile` with `encoding="none"` through this path would keep the large-file case covered.

File: cleanowners.py [hunk 3]
```diff
@@ -288,10 +289,10 @@ def get_codeowners_file(repo):
     codeowners_file_contents = None
     for path in (".github/CODEOWNERS", "CODEOWNERS", "docs/CODEOWNERS"):
         try:
-            codeowners_file_contents = repo.file_contents(path)
+            codeowners_file_contents = repo.get_contents(path)
             if codeowners_file_contents:
                 return codeowners_file_contents, path
-        except github3.exceptions.NotFoundError:
+        except GithubException:
```
Comment (zkoppert) [inline cleanowners.py]: `GithubException` is the base class of `UnknownObjectException` (404), `RateLimitExceededException` (403), and `BadCredentialsException` (401), so this now treats a rate-limit or auth failure the same as "file not found" and returns `(None, None)`. Downstream that counts the repo as missing a CODEOWNERS file, and in a non-dry-run it can open a spurious "add default CODEOWNERS" PR against a repo that already has one. I reproduced the `(None, None)` return with a mocked `RateLimitExceededException`. Catching the 404 subclass keeps the original 404-only semantics and lets real failures surface:

```suggestion
        except UnknownObjectException:
```

This needs `from github import GithubException, UnknownObjectException` at the top of the file.

File: cleanowners.py [hunk 4]
```diff
@@ -1,19 +1,20 @@
...
 from markdown_writer import write_step_summary, write_to_markdown
 
 
 def get_org(github_connection, organization):
     """Get the organization object"""
     try:
-        return github_connection.organization(organization)
-    except github3.exceptions.NotFoundError:
+        return github_connection.get_organization(organization)
+    except GithubException:
```
Comment (zkoppert) [inline cleanowners.py]: Same 404-broadening as the `get_codeowners_file` thread: catching the base `GithubException` here means a rate-limit or auth failure on the org lookup gets reported as "not an organization" and returns `None`. Catching the specific 404 keeps that from masking transient failures:

```suggestion
    except UnknownObjectException:
```

File: cleanowners.py [hunk 5]
```diff
@@ -164,19 +165,19 @@ def main():  # pragma: no cover
                     pull_count += 1
                     pull_request_urls.append(pull.html_url)
                     print(f"\tCreated pull request {pull.html_url}")
-                except github3.exceptions.NotFoundError:
+                except GithubException:
```
Comment (zkoppert) [inline cleanowners.py]: Non-blocking. Both PR-creation blocks broadened to `except GithubException`, so a 422 ("A pull request already exists"), a rate-limit, or a 5xx all print "Failed to create pull request. Check write permissions." and move on. Logging `exception.status` (or the message) here would make the real cause visible when a create fails for a reason other than permissions.

File: cleanowners.py (hunk 1 above)
Comment (jmeridth) [reply cleanowners.py]: @zkoppert Addressed in https://github.com/github-community-projects/cleanowners/commit/2ff90339fad634a4ef2c595b8248588f59264a6d. Resolved the username to a `NamedUser` via `github_connection.get_user(username)` before passing it to `has_in_members`.

File: cleanowners.py (hunk 2 above)
Comment (jmeridth) [reply cleanowners.py]: @zkoppert Addressed in https://github.com/github-community-projects/cleanowners/commit/2ff90339fad634a4ef2c595b8248588f59264a6d. Changed the guard to `encoding == "none"` and reused `codeowners_file_contents.sha` directly instead of a second `get_contents` call.

File: cleanowners.py (hunk 4 above)
Comment (jmeridth) [reply cleanowners.py]: @zkoppert Addressed in https://github.com/github-community-projects/cleanowners/commit/2ff90339fad634a4ef2c595b8248588f59264a6d. Narrowed to `UnknownObjectException` here as well.

File: cleanowners.py (hunk 3 above)
Comment (jmeridth) [reply cleanowners.py]: @zkoppert Addressed in https://github.com/github-community-projects/cleanowners/commit/2ff90339fad634a4ef2c595b8248588f59264a6d. Narrowed to `UnknownObjectException` to preserve 404-only semantics.

File: cleanowners.py (hunk 5 above)
Comment (jmeridth) [reply cleanowners.py]: @zkoppert Addressed in https://github.com/github-community-projects/cleanowners/commit/2ff90339fad634a4ef2c595b8248588f59264a6d. Both PR-creation except blocks now log `e.status` and `e.data` so the actual failure reason is visible.

File: cleanowners.py [hunk 6]
```diff
@@ -192,7 +195,8 @@ def main():  # pragma: no cover
                     break
 
                 # Check to see if the username is a member of the organization
-                if not gh_org.is_member(username):
+                member = github_connection.get_user(username)
+                if not gh_org.has_in_members(member):
```
Comment (zkoppert) [inline cleanowners.py]: Following up on my earlier `has_in_members` note (that thread went outdated once the code shifted): resolving the login to a `NamedUser` fixes the `AssertionError` for valid users, thanks. There's one edge case left, though, and it's the one this action exists to catch. `github_connection.get_user(username)` does a completed `GET /users/{username}` and raises `UnknownObjectException` on a 404, so a CODEOWNERS entry for a deleted or renamed individual account (a former employee, or a typo) will 404 here. That exception isn't caught locally, so it unwinds all the way to the outer `except Exception` that wraps the whole `for repo in repos` loop, prints `Error: 404 ...`, and skips every remaining repo. The action still exits 0, so it looks like a clean run while silently doing nothing further. github3's `is_member(username)` returned `False` for a nonexistent account (it hit the membership endpoint rather than checking user existence), so this is a behavior change from the port.

Wrapping the lookup and treating a missing account as a non-member keeps the run going and restores that graceful behavior (a deleted account is arguably the strongest signal to suggest removal):

```suggestion
                try:
                    member = github_connection.get_user(username)
                    is_member = gh_org.has_in_members(member)
                except UnknownObjectException:
                    # Account no longer exists, so treat it as a non-member to suggest removal
                    is_member = False
                if not is_member:
```

Since this path lives in `main()` (`# pragma: no cover`), a small test that drives a deleted-user handle through it would keep it from regressing.

File: cleanowners.py (hunk 1 above)
Comment (jmeridth) [reply cleanowners.py]: @zkoppert Addressed in https://github.com/github-community-projects/cleanowners/commit/50b8359. Wrapped the `get_user()` + `has_in_members()` call in a try/except for `UnknownObjectException`, treating deleted or renamed accounts as non-members so the run continues processing remaining repos.

File: cleanowners.py (hunk 2 above)
Comment (jmeridth) [reply cleanowners.py]: @copilot-pull-request-reviewer Addressed in https://github.com/github-community-projects/cleanowners/commit/50b8359. The `get_user()` call is now wrapped in a try/except for `UnknownObjectException`.

File: cleanowners.py (hunk 6 above)
Comment (jmeridth) [reply cleanowners.py]: @zkoppert Addressed in https://github.com/github-community-projects/cleanowners/commit/50b8359. Wrapped the `get_user()` + `has_in_members()` call in a try/except for `UnknownObjectException`, treating deleted or renamed accounts as non-members so the run continues processing remaining repos.

[the remark this answers is not in the document]
Comment (jmeridth) [reply cleanowners.py]: @copilot Addressed in https://github.com/github-community-projects/cleanowners/commit/50b8359. The `get_user()` call is now wrapped in a try/except for `UnknownObjectException`, treating unknown users as non-members.

File: auth.py [hunk 7]
```diff
@@ -10,7 +10,7 @@ def auth_to_github(
     gh_app_private_key_bytes: bytes,
```
[the remark this answers is not in the document]
Comment (jmeridth) [reply auth.py:10]: @copilot The `token` parameter is always a string. It comes from `os.environ.get("GH_TOKEN", "")` which defaults to an empty string, not `None`. The function guards on truthiness (`elif token:`) which correctly treats the empty string as "no token provided". The `str` annotation is accurate.

File: auth.py [hunk 8]
```diff
@@ -23,26 +23,22 @@ def auth_to_github(
...
-            gh = github3.github.GitHub()
-        gh.login_as_app_installation(
-            gh_app_private_key_bytes, str(gh_app_id), gh_app_installation_id
-        )
-        github_connection = gh
+            github_connection = Github(auth=installation_auth)
     elif ghe and token:
-        github_connection = github3.github.GitHubEnterprise(url=ghe, token=token)
+        github_connection = Github(base_url=f"{ghe}/api/v3", auth=Auth.Token(token))
     elif token:
```
[the remark this answers is not in the document]
Comment (jmeridth) [reply auth.py:38]: @copilot Addressed in https://github.com/github-community-projects/cleanowners/commit/50b8359. Added `ghe = ghe.rstrip("/")` before the URL construction to prevent double slashes.

File: test_auth.py [hunk 9]
```diff
@@ -11,15 +11,20 @@ class TestAuth(unittest.TestCase):
...
-        mock_login.return_value = "Authenticated to GitHub.com"
+        mock_token = MagicMock()
+        mock_auth.Token.return_value = mock_token
+        mock_github.return_value = "Authenticated to GitHub.com"
 
         result = auth.auth_to_github("token", "", "", b"", "", False)
 
+        mock_auth.Token.assert_called_once_with("token")
+        mock_github.assert_called_once_with(auth=mock_token)
         self.assertEqual(result, "Authenticated to GitHub.com")
```
[the remark this answers is not in the document]
Comment (jmeridth) [reply test_auth.py:28]: @copilot Addressed in https://github.com/github-community-projects/cleanowners/commit/50b8359. Changed empty strings to `None` for `gh_app_id` and `gh_app_installation_id` parameters.

File: test_auth.py [hunk 10]
```diff
@@ -35,73 +40,87 @@ def test_auth_to_github_without_token(self):
             "GH_TOKEN or the set of [GH_APP_ID, GH_APP_INSTALLATION_ID, GH_APP_PRIVATE_KEY] environment variables are not set",
         )
```
[the remark this answers is not in the document]
Comment (jmeridth) [reply test_auth.py:41]: @copilot Addressed in https://github.com/github-community-projects/cleanowners/commit/50b8359. Changed empty strings to `None`.

File: test_auth.py [hunk 11]
```diff
@@ -35,73 +40,87 @@ def test_auth_to_github_without_token(self):
...
         """
         Test the auth_to_github function when the GitHub Enterprise URL is provided.
         """
-        mock_ghe.return_value = "Authenticated to GitHub Enterprise"
+        mock_token = MagicMock()
+        mock_auth.Token.return_value = mock_token
+        mock_github.return_value = "Authenticated to GitHub Enterprise"
         result = auth.auth_to_github(
             "token", "", "", b"", "https://github.example.com", False
         )
```
[the remark this answers is not in the document]
Comment (jmeridth) [reply test_auth.py:55]: @copilot Addressed in https://github.com/github-community-projects/cleanowners/commit/50b8359. Changed empty strings to `None`.

File: test_auth.py [hunk 12]
```diff
@@ -35,73 +40,87 @@ def test_auth_to_github_without_token(self):
...
             "", "123", "123", b"123", "https://github.example.com", True
         )
-        mock.login_as_app_installation.assert_called_once_with(b"123", "123", "123")
-        self.assertEqual(result, mock)
 
-    @patch("github3.github.GitHub")
-    def test_auth_to_github_with_app(self, mock_gh):
+        mock_auth.AppAuth.assert_called_once_with(123, "123")
+        mock_app_auth.get_installation_auth.assert_called_once_with(123)
+        mock_github.assert_called_once_with(
```
[the remark this answers is not in the document]
Comment (jmeridth) [reply test_auth.py:80]: @copilot Addressed in https://github.com/github-community-projects/cleanowners/commit/50b8359. Changed string IDs to integers to match the `int | None` type annotation.

File: test_auth.py [hunk 13]
```diff
@@ -35,73 +40,87 @@ def test_auth_to_github_without_token(self):
...
-        mock.login_as_app_installation = MagicMock(return_value=True)
-        result = auth.auth_to_github("", 123, 456, b"private_key", "", False)
-        mock.login_as_app_installation.assert_called_once_with(
-            b"private_key", "123", 456
-        )
-        self.assertEqual(result, mock)
+        mock_auth.AppAuth.assert_called_once_with(123, "123")
+        mock_app_auth.get_installation_auth.assert_called_once_with(123)
+        mock_github.assert_called_once_with(auth=mock_installation_auth)
+        self.assertEqual(result, mock_github.return_value)
```
[the remark this answers is not in the document]
Comment (jmeridth) [reply test_auth.py:105]: @copilot Addressed in https://github.com/github-community-projects/cleanowners/commit/50b8359. Changed string IDs to integers.

**Review by zkoppert — APPROVED** [review]
🚀
